File: src/main/java/org/veupathdb/lib/gradle/container/util/JarFileFilter.java
package org.veupathdb.lib.gradle.container.util;

import org.gradle.api.Project;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.function.Function;
import java.util.stream.Stream;

/**
 * Methods for filtering files merged into fat jars.
 *
 * TODO: Add an option to add additional file filters.
 *
 * @since 2.0.0
 */
public class JarFileFilter {

  private static final FileFilter[] JarExclusionExtensions = {
    FileFilter.contains("log4j").and(FileFilter.contains(".dat")),
    FileFilter.endsWith(".sf"),
    FileFilter.endsWith(".dsa"),
    FileFilter.endsWith(".rsa"),
    FileFilter.endsWith(".md"),
  };

  @NotNull
  public static Function<File, Stream<Object>> expandFiles(@NotNull Project project) {
    return file -> {
      if (file.isDirectory())
        return Stream.of(file);

      return Stream.of(project.zipTree(file));
    };
  }

  public static boolean excludeJarFiles(@NotNull File file) {
    for (final var filter : JarExclusionExtensions) {
      if (!filter.test(file))
        return false;
    }

    return true;
  }
}

File: src/main/java/org/veupathdb/lib/gradle/container/util/JarFileFilter.kt
package org.veupathdb.lib.gradle.container.util

import org.gradle.api.file.FileTreeElement

/**
 * Methods for filtering files merged into fat jars.
 *
 * TODO: Add an option to add additional file filters.
 *
 * @since 2.0.0
 */
object JarFileFilter {

  @JvmStatic
  private val JarExclusionExtensions = arrayOf<(String) -> Boolean>(
    { it.contains("log4j") && it.contains(".dat") },
    { it.endsWith(".sf") },
    { it.endsWith(".dsa") },
    { it.endsWith(".rsa") },
    { it.endsWith(".md") },
  )

  @JvmStatic
  fun excludeJarFiles(file: FileTreeElement): Boolean {
    val name = file.name.toLowerCase()

    for (filter in JarExclusionExtensions) {
      if (filter(name))
        return false
    }

    return true
  }
}
